File: src/main/java/com/akagiyui/drive/component/RequestLogAspect.java
package com.akagiyui.drive.component;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import jakarta.servlet.http.HttpServletRequest;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.context.request.RequestAttributes;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.multipart.MultipartFile;

/**
 * 请求日志 切面
 *
 * @author AkagiYui
 */
@Aspect
@Component
@Slf4j
public class RequestLogAspect {

    private final ObjectMapper objectMapper = new ObjectMapper();

    public RequestLogAspect() {
        objectMapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY); // 可以序列化私有字段
        objectMapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false); // 序列化空对象时不抛异常
    }

    @Around("execution(* com.akagiyui.drive.controller.*.*(..))")
    public Object process(ProceedingJoinPoint joinPoint) throws Throwable {
        try {
            MethodSignature methodSignature = (MethodSignature)joinPoint.getSignature();  // 获取方法签名
            String methodName = joinPoint.getTarget().getClass().getName() + "." + methodSignature.getName();  // 方法名

            StringBuilder requestLog = new StringBuilder();
            HttpServletRequest request = getHttpServletRequest();

            // 获取客户端IP
            String clientIp = null;
            if (request != null) {
                clientIp = request.getHeader("X-Real-IP");
            }
            if (!StringUtils.hasText(clientIp) && request != null) {
                clientIp = request.getRemoteAddr();
            }
            if (request != null) {
                requestLog.append(String.format("%n %s request -> %s", clientIp, methodName));
            }

            // 获取请求参数
            Object[] args = joinPoint.getArgs();  // 方法参数
            if (args.length > 0) {
                StringBuilder paramsBuffer = new StringBuilder();
                for (Object param : args) {
                    if (param instanceof MultipartFile) {
                        continue; // 忽略文件参数
                    }
                    String paramStr = objectMapper.writeValueAsString(param);
                    if (paramStr.length() > 1000) {
                        paramStr = paramStr.substring(0, 1000) + "...";
                    }
                    paramsBuffer.append(paramStr).append(", ");
                }
                if (!paramsBuffer.isEmpty()) {
                    String paramStr =paramsBuffer.substring(0, paramsBuffer.length() - 2);
                    if (paramStr.length() > 1000) {
                        paramStr = paramStr.substring(0, 1000) + "...";
                    }
                    requestLog.append(String.format("%n params -> %s", paramStr));
                }
            }

            // 记录返回值或异常
            long startTime = System.currentTimeMillis();
            try {
                Object returnObj = joinPoint.proceed();
                if (returnObj != null) {
                    String returnStr = objectMapper.writeValueAsString(returnObj);
                    if (returnStr.length() > 1000) {
                        returnStr = returnStr.substring(0, 1000) + "...";
                    }
                    requestLog.append(String.format("%n result -> %s", returnStr));
                }
                return returnObj;
            } catch (Throwable throwable) {
                requestLog.append(String.format("%n error -> %s", throwable.getMessage()));
                throw new ServiceThrowable(throwable.getMessage(), throwable);
            } finally {
                long endTime = System.currentTimeMillis();
                requestLog.append(String.format("%n time -> %dms", endTime - startTime));
                log.debug(requestLog.toString());
            }

        } catch (ServiceThrowable serviceThrowable) {
            throw serviceThrowable.getCause();
        } catch (Throwable throwable) {
            log.error("RequestLog AOP error", throwable);
            return joinPoint.proceed();
        }
    }

    private HttpServletRequest getHttpServletRequest() {
        RequestAttributes requestAttributes = RequestContextHolder.getRequestAttributes();
        if (requestAttributes == null) {
            return null;
        }
        return (HttpServletRequest)requestAttributes.resolveReference(RequestAttributes.REFERENCE_REQUEST);
    }

    @Getter
    static class ServiceThrowable extends Throwable {
        private final Throwable cause;

        public ServiceThrowable(String message, Throwable cause) {
            super(message);
            this.cause = cause;
        }
    }
}

File: app/src/main/kotlin/com/akagiyui/drive/component/RequestLogAspect.kt
package com.akagiyui.drive.component

import com.akagiyui.common.delegate.LoggerDelegate
import com.akagiyui.common.utils.compressPackageName
import com.akagiyui.common.utils.ellipsis
import com.akagiyui.common.utils.hasText
import com.akagiyui.common.utils.toStr
import com.akagiyui.drive.entity.User
import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.PropertyAccessor
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import io.minio.GetObjectResponse
import jakarta.servlet.http.HttpServletRequest
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import org.springframework.web.context.request.RequestAttributes
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter

/**
 * 请求日志 切面
 *
 * @author AkagiYui
 */
@Aspect
@Component
class RequestLogAspect {

    private val log by LoggerDelegate()
    private val objectMapper = ObjectMapper().apply {
        setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY) // 可以序列化私有字段
        configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false) // 序列化空对象时不抛异常
        registerKotlinModule() // 添加 Kotlin 支持
    }

    @Around("execution(@(org.springframework.web.bind.annotation.*Mapping) * *(..))")
    @Throws(Throwable::class)
    fun process(joinPoint: ProceedingJoinPoint): Any? {

        val methodSignature = joinPoint.signature as MethodSignature // 获取方法签名
        val methodName = "${joinPoint.target::class.qualifiedName}::${methodSignature.name}" // 方法名

        val requestLog = StringBuilder()
        val request = httpServletRequest ?: return joinPoint.proceed()

        val clientIp: String = request.getHeader("X-Real-IP") ?: request.remoteAddr // 客户端IP
        var requestLine = "\n $clientIp -> [${request.method}]" // HTTP请求方法
        request.contentType.hasText { requestLine += "($it)" } // 请求类型
        requestLine += request.requestURI // 请求路径
        requestLog.append(requestLine)

        // 请求params
        request.parameterMap.also { params ->
            val paramsBuffer: String = params
                .filterNot { it.key == "password" }
                .map { (key, value) -> "$key=${value.joinToString(",")}".ellipsis(100) }
                .joinToString(", ")
                .ellipsis(1000)
            requestLog.append(if (paramsBuffer.isNotBlank()) "($paramsBuffer)" else "")
        }

        // 获取方法参数
        requestLog.append("\n method -> ${methodName.compressPackageName(50)}")
        val args = joinPoint.args // 方法参数
        val paramsBuffer: String = args
            .asSequence()
            .filterNotNull()
            .map { anyToString(it).ellipsis(100) }
            .joinToString(", ") // 将参数连接成字符串
            .ellipsis(1000)
        requestLog.append("($paramsBuffer)")

        // 记录返回值与异常
        return try {
            val startTime = System.currentTimeMillis()
            joinPoint.proceed().also { returnObj ->
                val duration = System.currentTimeMillis() - startTime
                val resultLog = anyToString(returnObj).ellipsis(500)
                requestLog.append("\n result[${duration}ms] <- $resultLog")
            }
        } catch (throwable: Throwable) {
            requestLog.append("\n error <- ${throwable.message}")
            throw throwable
        } finally {
            log.debug(requestLog.toString())
        }
    }

    private fun anyToString(any: Any?): String {
        if (any == null) {
            return "null"
        }
        if (any::class.isData) {
            return any.toString()
        }
        if (any is ByteArray) {
            return any.toStr()
        }
        if (any is GetObjectResponse) {
            return any.toStr()
        }
        if (any is ResponseEntity<*>) {
            return any.toStr()
        }
        if (any is MultipartFile) {
            return "MultipartFile[${any.originalFilename}]"
        }
        if ((any is Collection<*>) && any.isNotEmpty() && (any.first() is MultipartFile)) {
            return "Collection<MultipartFile>[${any.size}]"
        }
        if (any is SseEmitter) {
            return any.toString()
        }
        if (any is User) {
            return "User[${any.id}]"
        }
        return objectMapper.writeValueAsString(any)
    }

    private val httpServletRequest: HttpServletRequest?
        get() {
            val requestAttributes = RequestContextHolder.getRequestAttributes() ?: return null
            return requestAttributes.resolveReference(RequestAttributes.REFERENCE_REQUEST) as HttpServletRequest?
        }

}
